imports/plugins/core/catalog/server/methods/catalog.js

```diff
@@ -0,0 +1,241 @@
+  product.isLowQuantity = isLowQuantity(variants);
+  product.variants = variants.map((variant) => {
+    const { inventoryQuantity, ...v } = variant;
+    return v;
+  });
+
+  // Insert/update catalog document
+  const result = CatalogCollection.upsert({
+    _id: productId
+  }, {
+    $set: product
+  });
+
+  return result && result.numberAffected === 1;
+}
+
+/**
+ * @method publishProductsToCatalog
+ * @summary Publish one or more products to the Catalog
+ * @memberof Catalog
+ * @param {string|array} productIds - A string product id or an array of product ids
+ * @return {boolean} true on successful publish for all documents, false if one ore more fail to publish
+ */
+export function publishProductsToCatalog(productIds) {
+  check(productIds, Match.OneOf(String, Array));
+
+  let ids = productIds;
+  if (typeof ids === "string") {
+    ids = [productIds];
+  }
+
+  return ids.every(async (productId) => await publishProductToCatalog(productId));
+}
+
+/**
+ * @method publishProductInventoryAdjustments
+ * @summary Publish inventory updates for a single product to the Catalog
+ * @memberof Catalog
+ * @param {string} productId - A string product id
+ * @return {boolean} true on success, false on failure
+ */
+export function publishProductInventoryAdjustments(productId) {
+  check(productId, Match.OneOf(String, Array));
+
+  const catalogProduct = CatalogCollection.findOne({
+    _id: productId
+  });
+
+  if (!catalogProduct) {
+    Logger.info("Cannot publish inventory changes to catalog product");
+    return false;
+  }
+
+  const variants = Products.find({
+    ancestors: {
+      $in: [productId]
+    }
+  }).fetch();
+
+  const update = {
+    isSoldOut: isSoldOut(variants),
+    isBackorder: isBackorder(variants),
+    isLowQuantity: isLowQuantity(variants)
+  };
+
+  // Only apply changes of one these fields have changed
+  if (
+    update.isSoldOut !== catalogProduct.isSoldOut ||
+    update.isBackorder !== catalogProduct.isBackorder ||
+    update.isLowQuantity !== catalogProduct.isLowQuantity
+  ) {
+    const result = CatalogCollection.update({
+      _id: productId
+    }, {
+      $set: update
+    });
+
+    return result;
+  }
+
+  return false;
+}
+
+Meteor.methods({
+  "catalog/publish/products": (productIds) => {
+    check(productIds, Match.OneOf(String, Array));
+
+    // Ensure user has createProduct permission for active shop
+    if (!Reaction.hasPermission("createProduct")) {
+      Logger.error("Access Denied");
+      throw new Meteor.Error("access-denied", "Access Denied");
+    }
+
+    // Convert productIds if it's a string
+    let ids = productIds;
+    if (typeof ids === "string") {
+      ids = [productIds];
+    }
+
+    // Find all products
+    const productsToPublish = Products.find({
+      _id: { $in: ids }
+    }).fetch();
+
+    if (Array.isArray(productsToPublish)) {
+      const canUpdatePrimaryShopProducts = Reaction.hasPermission("createProduct", this.userId, Reaction.getPrimaryShopId());
```

Comment:
Why not do this check at the start of the method with the other permission check?